### ISSUE TITLE
Nvidia Gamescope VRR (Monitor VRR reverting to max instead of staying in range) on Wayland

### ISSUE THREAD

**Issue by jrgiacone** [ValveSoftware/gamescope#1617]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Are you using any gamescope patches or a forked version of gamescope?

- [X] The issue occurs on upstream gamescope without any modifications

### Current Behavior

I am currently using Arch with Plasma on Wayland and Nvidia 565. Unsure if it is a gamescope issue or a nvidia driver issue. However, when using gamescope and turning on the monitors frame rate HUD to see the current refresh rate with a game like new world running in gamescope with HDR on and FPS in the 90-100 range the monitor will consistently jump to the max refreshrate. When gamescope is not used the monitor matches and stays within the 90-100 range without bouncing to max refresh rate.

### Steps To Reproduce

1. Launch New world or any game with HDR with the following options: `gamescope -W 3440 -H 1440 -r 165 --hdr-enabled --force-grab-cursor -f --mangoapp -- gamemoderun %command%`
2. Turn on monitor HUD to see current refresh rate. I am using a aw3423DWF and a LG27gl83a-b and it is present on both.
3. While gaming play a game that keeps the FPS below the max monitor refresh rate, ideally 90-120 of 165 where you know it will not jump to 165. Watch as the monitor continually shoots past to max refresh rate while the fps does not dip below 90 or 60 on 1% lows.
4. Run the same test without gamescope and just gamemoderun '%command%' in steam and the issue is not present.

### Hardware information

```markdown
- Distro: Arch
- CPU: amd 5800x
- GPU: Nvidia 3080
- Driver Version: 565.57.01-2
```

### Software information

```markdown
- Desktop environment: KDE 6.2.3
- Session type: Wayland
- Gamescope version: 	3.15.14-1
- Gamescope launch command(s): `gamescope -W 3440 -H 1440 -r 165 --hdr-enabled --force-grab-cursor -f --mangoapp -- gamemoderun %command%`
```

### Which gamescope backends have the issue you are reporting?

- [X] Wayland (default for nested gamescope)
- [ ] DRM (default for embedded gamescope, i.e. gamescope-session)
- [ ] SDL
- [ ] OpenVR

**Comment by matte-schwartz** [comment]
what about if you drop off the `--mangoapp` argument and just check your monitor's built-in refresh rate OSD? 

there is a known issue within gamescope where mangohud overlays can spike your monitor's refresh rate up to maximum in both the Wayland backend and the DRM backend, especially ~60% or below FPS of your monitor's true refresh rate

**Comment by jrgiacone** [comment]
> what about if you drop off the `--mangoapp` argument and just check your monitor's built-in refresh rate OSD?
> 
> there is a known issue within gamescope (and not GPU specific) where mangohud overlays can spike your monitor's refresh rate up to maximum in both the Wayland backend and the DRM backend, especially ~60% or below FPS of your monitor's true refresh rate

I will try removing mango app when I get in this evening! I don't think I noticed with global mango hud running via goverlay, but yea i used mangoapp to validate HDR was on

**Comment by jrgiacone** [comment]
> what about if you drop off the `--mangoapp` argument and just check your monitor's built-in refresh rate OSD?
> 
> there is a known issue within gamescope (and not GPU specific) where mangohud overlays can spike your monitor's refresh rate up to maximum in both the Wayland backend and the DRM backend, especially ~60% or below FPS of your monitor's true maximum refresh rate.
> 
> this can also be triggered by: a cursor, GamepadUI side menus, and toast notifications from Steam

I've tried removing everything up to just `gamescope -H 3440 -W 1440 -- gamemoderun %command%` and it is the same issue. It is even notable just in the menu and I could record it to show. In the menu when launched with gamescope the VRR fluctuates from 80-165 when FPS are at 110.

When I just run gamemoderun %command% it holds at 104 as shown in the menu.

Would a series of videos showing it help?

**Comment by jrgiacone** [comment]
https://github.com/user-attachments/assets/c73a4d6b-5610-4404-b6d0-f798b53c8c77


https://github.com/user-attachments/assets/2e47d0dd-fa3d-40e9-b182-aac3b07f91ad

I've attached both videos we will see how they play, but in one you can clearly see the impact of gamescope

**Comment by matte-schwartz** [comment]
yeah it looks exactly the same issue as the DRM backend, but with the DRM backend you can at least mitigate it.

I have an open bug report for it here if you want to leave a comment on it: https://bugs.kde.org/show_bug.cgi?id=493090

**Comment by matte-schwartz** [comment]
@ammuench what GPU do you have? would be helpful to know more details about your own VRR issue as well.

**Comment by ammuench** [comment]
> @ammuench what GPU do you have? would be helpful to know more details about your own VRR issue as well.

Using an NVIDIA 4070ti on latest Arch Drivers at time of writing (565.77-5)

**Comment by matte-schwartz** [comment]
Okay thanks for confirming.  As noted above, I've seen the same issue with Quake II RTX in both Plasma 6 Wayland and Cosmic on a 7900XTX, so it's not a NVIDIA-specific issue. Previously I had only tested this behavior on Plasma which is why I filed that bug report I linked above, but clearly it can happen with other Wayland compositors as well so that was the wrong place to pursue this.

If you don't mind one more experiment, if you launch Steam embedded in gamescope from a TTY (like `gamescope -e -h 1440 -w 3440 -r 175 --adaptive-sync -- steam -steamdeck` for example) and check the same games, does the monitor similarly spike to maximum refresh rate or does it behave properly? Make sure to remove the gamescope launch option from within the game's settings so you don't run double gamescope. when you're done, you can switch to a different TTY session and run `gamescopectl shutdown` to close gamescope/steam

**Comment by ammuench** [comment]
It may take me a couple days to sit down and try proper but I'll give it a go soon and report back

**Comment by njdom24** [comment]
I can confirm I'm seeing this issue on an AMD RX 5700 XT with the embedded (`drm`) via TTY, and nested (`wayland`, `sdl`) backends in SwayWM (which works well with VRR on its own).

**Comment by njdom24** [comment]
https://github.com/user-attachments/assets/be4f77c7-1bb3-42ac-b461-219f03f1935d

FWIW, my monitor's FPS counter updates in realtime

**Comment by matte-schwartz** [comment]
@njdom24 you're confusing me now... where is this video clip from and with what launch commands are you using? 

for now, lets keep the two backends separate imo. the issue as reported here should only be for the nested VRR issue and https://github.com/ValveSoftware/gamescope/issues/1513 should be for situations where VRR spikes in the drm backend, unless I see some more convincing evidence that they are actually the exact same issue.

For example, I can reproduce this specific problematic VRR spiking behavior in Quake II RTX, Ghost of Tsushima, and Indiana Jones and the Great Circle **only** while using gamescope nested on Plasma 6 Wayland and Cosmic desktop, while I cannot reproduce it in the drm backend unless I start messing with overlays, etc (which leads to https://github.com/ValveSoftware/gamescope/issues/1513)

**Comment by njdom24** [comment]
@matte-schwartz Sorry. The clip is with the embedded backend with `MANGOHUD=0 gamescope -e -O DP-1 -H 1440 -r 144 --adaptive-sync -- steam -tenfoot -pipewire-dmabuf -console -cef-force-gpu`

**Comment by matte-schwartz** [comment]
which game? if I have it, I'll try to check and see if it's the same underlying issue between both backends on my own setup to hopefully clear up some confusion here

**Comment by njdom24** [comment]
Elden Ring. I'll test a few others I have installed over the next few days.

**Comment by matte-schwartz** [comment]
fwiw Elden Ring looks similarly jumpy to me between Gamescope in the DRM backend and KDE VRR without gamescope at all, so it may be a bad game to test. for the future lets keep this thread for the nested side of things and the other thread for embedded  to keep things organized, thanks.

**Comment by njdom24** [comment]
Good call on Quake II RTX. It's perfect with the DRM backend, but flickering like mad nested in Sway.

**Comment by jrgiacone** [comment]
do we think it is a nvidia issue or a gamescope/kde issue?

**Comment by matte-schwartz** [comment]
It's a Gamescope issue when nested on another Wayland compositor, the same issue occurs on other DEs (I checked cosmic-comp w/ VRR as well) and GPU manufacturers

**Comment by SeimusS** [comment]
I came across similar (maybe same?) problem. But initially I thought its a problem of HDR so I opened a BUG to KDE, where I was guided that this is most likely a gamescope issue and found this topic.

https://bugs.kde.org/show_bug.cgi?id=502202

I set the parameter:
gamescope -w 3440 -h 1440 -f -r 165 --adaptive-sync --hdr-enabled --  %command%

= No flicker

I tried as well same parameter with using gooverlay:
gamescope -w 3440 -h 1440 -f -r 165 --adaptive-sync --hdr-enabled --  %command%

= No flicker

I tried as well to set mangoapp as a parameter
gamescope -w 3440 -h 1440 -f -r 165 --adaptive-sync --hdr-enabled --mangoapp --  %command%

= No flicker

And the flickering is gone. So looks like in my case this was a user mistake. If gamesscope is used with a VRR monitor e.g Adaptive sync is turned on, gamescope has to have this parameter explicitly set otherwise VRR is broken and causes insane flicker bellow ~100FPS.

I hope my case is matching this problem, if not sorry for the confusion.

Regards,
S.

**Comment by x5f3759df** [comment]
Have same issue on NVIDIA card and 575 beta - without gamescope everything is fine, but with it I have VRR fluctuation like OP does.

**Comment by OlegAckbar** [comment]
I can confirm this issue on a Gamescope 3.16.4 with AMD 6700 XT. Disabling mangoapp doesn't help.

https://github.com/user-attachments/assets/f41056b3-9ec1-439f-aeda-4b9831a7c548 

https://github.com/user-attachments/assets/05f5c83a-2581-4d44-829a-209ae5f2e966

**Comment by matte-schwartz** [comment]
As of Gamescope 3.16.5 I can't reproduce this issue anymore on AMD. 3.16.4 was the last broken version when I was testing today, and 3.16.5 was the first good version. 

Tested Gamescope nested on both KWin and Cosmic, and both worked.

**Comment by RevengeRip** [comment]
@matte-schwartz 3.16.5? Are you sure? I can still reproduce it on the latest git.

**Comment by matte-schwartz** [comment]
Which game(s)? I tried for a while yesterday with Quake 2 RTX and Ghost of Tsushima since I could previously reproduce the issue with those, but I wasn't able to trigger it at all anymore with `gamescope -f -h 2160 -w 5120 --adaptive-sync -- %command%`. I'll re-check later, maybe there's something else I'm missing.

**Comment by RevengeRip** [comment]
@matte-schwartz huh, I didn't know about `--adaptive-sync`. With it there are no issues. Although some games are crashing when using it. I think I should file a relevant bug report.

**Comment by matte-schwartz** [comment]
Yes, please do and mention which games. Hopefully I have at least one of them.

**Comment by RevengeRip** [comment]
> Yes, please do and mention which games. Hopefully I have at least one of them.

https://github.com/ValveSoftware/gamescope/issues/1915

**Comment by njdom24** [comment]
I can _technically kinda_ reproduce this issue by setting an FPS limit and using MangoHud's `fps_limit_method=early`. Even `mangoapp`'s frametime graph is perfectly flat, but VRR spikes like mad. `late` doesn't have this problem, nor does `early` outside of gamescope, which is a bit of a shame, since `early` is the smoothest FPS limiter I've used.

It also seems fine in the embedded backend, and the exclusively on the Wayland backend under Sway with `output "*" max_render_time 1`. Plasma and default Sway present the issue.

**Comment by matte-schwartz** [comment]
there are enough differentiating factors there that it probably warrants a separate report @njdom24

**Comment by x5f3759df** [comment]
This issue persists and fixed by using `gamescopectl adaptive_sync_ignore_overlay`

But I don't know if gamescope even has options do to it permanently

**Comment by njdom24** [comment]
@x5f3759df What version of Gamescope are you running? I haven't been seeing this issue on KDE and Sway since 3.16.11.

**Comment by neon-grim** [comment]
> reply to https://github.com/ValveSoftware/gamescope/issues/1617#issuecomment-3201514667

I'm still having flicker issues when running gamescope through hyprland in nested mode and enabling VRR. Refresh rate counter behaves erratic when playing with VRR active.
